### ISSUE TITLE
Preferences Sync making too many requests

### ISSUE THREAD

**Issue by CleanCut** [microsoft/vscode#98625]
Issue Type: <b>Bug</b>

A dialog popped up and said preferences sync was being disabled because of too many requests...and it told me to submit an issue. So here I am!

VS Code version: Code - Insiders 1.46.0-insider (b26e0bcf39c1b8faaf770265391b5a387d1d9172, 2020-05-26T17:07:53.549Z)
OS version: Darwin x64 19.5.0

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Xeon(R) W-3245 CPU @ 3.20GHz (32 x 3200)|
|GPU Status|2d_canvas: enabled<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: enabled<br>metal: disabled_off<br>multiple_raster_threads: enabled_on<br>oop_rasterization: disabled_off<br>opengl: enabled_on<br>protected_video_decode: unavailable_off<br>rasterization: enabled<br>skia_renderer: disabled_off_ok<br>video_decode: enabled<br>viz_display_compositor: enabled_on<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|2, 3, 2|
|Memory (System)|192.00GB (152.27GB free)|
|Process Argv|-psn_0_393312|
|Screen Reader|no|
|VM|0%|
</details><details><summary>Extensions (7)</summary>

Extension|Author (truncated)|Version
---|---|---
vscode-ron|a5h|0.6.0
better-toml|bun|0.3.2
rust-analyzer|mat|0.2.182
python|ms-|2020.5.80290
vscode-yaml|red|0.8.0
crates|ser|0.4.9
vscode-lldb|vad|1.5.0

</details>

**Comment by CleanCut** [comment]
Yep, looks like lots of duplicates. Dunno which one to close in favor of.

**Comment by CleanCut** [comment]
Here's the [userDataSync.log](https://github.com/microsoft/vscode/files/4686456/userDataSync.log) for my session.

**Comment by sandy081** [comment]
@CleanCut, @manovotny, @texnicii, @FizzBuzz791

 Can you please let us know which account are you using to sync (GitHub or Microsoft) ?

**Comment by FizzBuzz791** [comment]
GitHub account

**Comment by texnicii** [comment]
GitHub too

**Comment by sandy081** [comment]
Thanks @CleanCut for the logs.

I see there is a loop in auto sync that is causing this. I reverse engineered the logs and I am suspecting the loop is caused by following

>[2020-05-26 22:22:57.973] [userDataSync] [info] Auto Sync: Triggered by Interval
[2020-05-26 22:23:04.425] [userDataSync] [info] Settings: No changes found during synchronizing settings.
[2020-05-26 22:23:04.428] [userDataSync] [info] Keybindings: No changes found during synchronizing keybindings.
[2020-05-26 22:23:04.430] [userDataSync] [info] Snippets: No changes found during synchronizing snippets.
[2020-05-26 22:23:04.430] [userDataSync] [info] GlobalState: Skipped synchronizing globalstate as it is disabled.
[2020-05-26 22:23:04.483] [userDataSync] [info] Extensions: No changes found during synchronizing extensions.
[2020-05-26 22:23:04.483] [userDataSync] [info] Sync done. Took 2268ms 

Sync done successfully.

> [2020-05-26 22:28:04.477] [userDataSync] [info] Auto Sync: Triggered by Interval

Auto sync is triggered by 5 min interval 

> [2020-05-26 22:28:05.474] [userDataSync] [info] Auto Sync: Stopped

Sync is not successful and auto sync is disabled/stopped after a second. This could be caused by Unauthorised failure (`401`) during which we reset the token and request auth extension (Microsoft, GitHub) with refreshed sessions (tokens)

> [2020-05-26 22:28:05.477] [userDataSync] [info] Auto Sync: Started
[2020-05-26 22:28:05.477] [userDataSync] [info] Auto Sync: Triggered by Interval

We received active refreshed sessions from the extension and updated the token which triggered auto sync

> [2020-05-26 22:28:05.761] [userDataSync] [info] Auto Sync: Stopped

Seems token is not valid and Auto Sync is stopped and the loop `Invalid token -> Reset token -> Stop sync -> Update token -> Start sync -> Invalid token`

@RMacfarlane I suspect this is happening only for **GitHub** accounts because for **Microsoft** accounts you are refreshing the token when asked for sessions 

https://github.com/microsoft/vscode/blob/87b0c54586a5f0d39fb04d5dad9fbfe96f7d677c/extensions/microsoft-authentication/src/AADHelper.ts#L242-L244

https://github.com/microsoft/vscode/blob/87b0c54586a5f0d39fb04d5dad9fbfe96f7d677c/extensions/microsoft-authentication/src/AADHelper.ts#L207-L231

But for **GitHub** accounts, I do not see sessions/tokens are validated/refreshed when asked for 

https://github.com/microsoft/vscode/blob/87b0c54586a5f0d39fb04d5dad9fbfe96f7d677c/extensions/github-authentication/src/github.ts#L142-L144

**Comment by sandy081** [comment]
To continue...

If I try to turn on sync after above issue, I am getting following error

<img width="772" alt="image" src="https://user-images.githubusercontent.com/10746682/83014968-f8fd4680-a01f-11ea-828c-9e8049b995a6.png">

and I see in my network log that authorization header is being sent with token.

**Comment by olofwalker** [comment]
I see the same issue as @sandy081, no longer able to sync settings.

**Comment by sandy081** [comment]
@Tyriar FYI

**Comment by sandy081** [comment]
I signed out of my GitHub account and tried to turn on sync by freshly signing in which also did not work. I can confirm that I am sending a fresh different token.

@Tyriar May I know if this is getting handled correctly on server?

**Comment by sandy081** [comment]
Please try to turn on sync now and let us know if it works

**Comment by Tyriar** [comment]
I reverted the latest deployment so this is working again. I haven't root caused this yet but we need to add some protections in the client so that something like this doesn't turn off settings sync next time; if the server says a token is unauthorized, the client should be using exponential backoff so that the client-side request limit isn't hit.

**Comment by sandy081** [comment]
> if the server says a token is unauthorized, the client should be using exponential backoff so that the client-side request limit isn't hit.

Client is doing the right thing here by disabling the auto sync when the server returns `401` and enable auto sync only after the token is available.

**Comment by Tyriar** [comment]
@sandy081 retrying a failed request every 300ms is not the right thing. Unauthorized from the server should not trigger client-side rate limiting.

**Comment by sandy081** [comment]
@Tyriar It is not retrying the failed request, it is restarting after getting the valid token. How can the client knows if the token is valid or not. Adding exponential backoff in this case is jus to covers up the bug that is caused some where.

**Comment by manovotny** [comment]
@sandy081 GitHub account for me as well.

**Comment by Tyriar** [comment]
@sandy081 not having exponential backoff causes clients to hammer the server and then turn off the client. We never want sync to turn off unless there is something really wrong.

**Comment by sandy081** [comment]
@Tyriar We do exponential backoff for all errors. For `401` we are disabling the sync completely until we get a valid token. If we also backoff for this then it will impact positive scenarios - to wait minutes before sync gets kicked off and we degrade experience for majority. 

I agree this is a bad bug and its good that this got detected by the protective mechanism.

**Comment by sandy081** [comment]
Closing this as @Tyriar pushed the fix on the server.

**Comment by CleanCut** [comment]
@sandy081 I was also using my GitHub account.

I just updated and re-enabled sync. It seems to be working so far.

**Comment by ssbarnea** [comment]
I can confirm that after the service side changes where made, it started to work.

**Comment by rosstex** [comment]
Still getting this issue.

**Comment by Chazza** [comment]
I get this issue on a fresh install.

**Comment by sandy081** [comment]
Please check if you are on latest insiders and if so and if you still see this, file a new issue.
